Update Azure Monitor Exporter to use the new internal traces format

pcwiese opened the pull request:
**Description:**
This change brings the Azure Monitor trace exporter up to date with
the latest OTEL attribute specifications for HTTP, RPC, Database, and
Messaging Spans. It does not implement any mapping yet for FaaS Spans.

This is a complete refactor to reduce code duplication and increase readability.

**Testing:** 
- [X] I ran Unit Tests locally.
- [X] Tested E2E using a recent version of the OpenTelemetry dotnet SDK with matching otel proto files emitting HTTP server and client Spans.

**Documentation:**
An update to the README will follow in a separate PR from a colleague.

pcwiese commented:
@youngbupark

pcwiese commented on exporter/azuremonitorexporter/conventions.go:
```diff
@@ -0,0 +1,258 @@
+)
+
+/*
+	This file encapsulates the extraction logic for the various kinds attributes
```
This entire file could be moved and reused by other trace exporters potentially

frigus02 commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+		data.BaseData = remoteDependencyData
+		data.BaseType = remoteDependencyData.BaseType()
+	} else if spanKind == pdata.SpanKindINTERNAL {
+		messageData := spanToMessageData(span, incomingSpanType)
```
I saw that both Azure Monitor exporter for Python (see [here](https://github.com/microsoft/opentelemetry-azure-monitor-python/blob/e9e669f51696d081f9816e4121f269a124fef6a2/azure_monitor/src/azure_monitor/export/trace/__init__.py#L143)) and JavaScript (see [here](https://github.com/microsoft/opentelemetry-azure-monitor-js/blob/295db7379cb504616be6aedd1983aee88e8bb7c6/packages/azure-opentelemetry-exporter/src/utils/spanUtils.ts#L193)) map internal spans to a remote dependency data with the type `InProc`. This made sense to me because message data doesn't have an ID and so it can't have children.

In my mind I've always mapped span events to message data instead.

However I am quite new to Azure Monitor, so I may be misunderstanding some things.

pcwiese replied:
An excellent point! @youngbupark suggested MessageData but I think I will change this back to InProc. I believe it was like that before.

youngbupark replied:
Good point. I also found https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python-dependency#in-process-dependencies . The original code seems correct. apologies for wrong info.

pcwiese replied:
I totally agree with this and I'll make the change. I think it is InProc in the current version...

frigus02 reviewed (commented):
I just started looking into updating the Azure Monitor exporter. This PR comes at a great time. Thanks a lot.

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+	}
+
+	// Database
+	if _, exists := attributeMap.Get(conventions.AttributeDBType); exists {
```
in the latest spec, `db.type` is renamed to `db.system` - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#connection-level-attributes

pcwiese replied:
yeah looks like the spec was revised 2 days ago. I'll make this change.

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+
+	data.Type = attrs.DBType
+	data.Data = attrs.DBStatement
+	data.Target = attrs.DBURL
```
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#call-level-attributes

* data.Type : DBType seems deprecated. need to switch to `db.system`
* data.Data : we need to look at `db.operation` as a fallback.
* data.Target: `db.url` is deprecated so that we need to use net.peer.*
for example, data.Target will be
1. `net.peer.name:net.peer.port`
2. `net.peer.ip:net.peer.port`
3. `net.peer.name`
4. `net.peer.ip`

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+	// TODO Understand how to map attributes to RemoteDependencyData fields
+	data.Data = attrs.MessagingURL
+	data.Type = attrs.MessagingSystem
+	data.Target = attrs.MessagingDestination
```
I have thought that we always need to look at `net.peer.*` as a fallback for all types.  How do you think?

pcwiese replied:
I agree with this also.

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+	data := contracts.NewRemoteDependencyData()
+	data.Id = idToHex(span.SpanID())
+	data.Name = span.Name()
+	data.ResultCode, data.Success = getDefaultFormattedSpanStatus(span.Status())
```
👍  this makes sense. The previous code didn't look at span status by default.

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+
+func getFormattedHTTPStatusValues(statusCode int64) (statusAsString string, success bool) {
+	if statusCode != 0 {
+		return strconv.FormatInt(statusCode, 10), statusCode <= 399
```
`statuscode <= 299` is correct.

I also would check lower bound as well. statuscode >= 200 && statuscode <= 299

pcwiese replied:
right.

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+func fillRequestDataHTTP(span pdata.Span, data *contracts.RequestData) {
+	attrs := copyAndExtractHTTPAttributes(span.Attributes(), data.Properties, data.Measurements)
+
+	data.ResponseCode, data.Success = getFormattedHTTPStatusValues(attrs.HTTPStatusCode)
```
What if user set only span.status without http_status_code attribute? then it will always be 200 success. can we take this attribute only if this attr exists?

pcwiese replied:
sure. I just won't touch data.ResponseCode or data.Success if the attribute isn't set

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,660 @@
+	attrs := copyAndExtractHTTPAttributes(span.Attributes(), data.Properties, data.Measurements)
+
+	data.Type = "HTTP"
+	data.ResultCode, data.Success = getFormattedHTTPStatusValues(attrs.HTTPStatusCode)
```
like above, we may need to consider the case when users set only span.status.

youngbupark reviewed (commented):
Thanks for this work :) I added some comments. overall looks great.

pcwiese commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,683 @@
+		}
+
+		if instrumentationLibrary.Version() != "" {
+			dataProperties["instrumentationlibrary.name"] = instrumentationLibrary.Version()
```
good catch. I need to add these to a test. I'll add them both as constants

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,679 @@
+	data.Type = attrs.MessagingSystem
+
+	if attrs.MessagingURL != "" {
+		data.Target = attrs.MessagingURL
```
message.destination is `Required` field on the spec, so referring to `messaging.destination` also makes sense to me. (I saw your previous pr referred to this attr.) how do you think?

1. messaging.url
2. messaging.destination
3. net.peer.*

![image](https://user-images.githubusercontent.com/11863108/86535171-32cd3280-be93-11ea-9e27-49b117490c54.png)

pcwiese replied:
makes sense. I did have it like that previously but seemed a little redundant because it also the Span name probably.

I’ll fix the tests also

youngbupark replied:
I just did some validations on appinsight. 

Let's assume that single publisher and multiple subscribers for topicA are tracked by the same appinsight and show the links between single source to multiple targets.

But publisher linked to one of them because Target might be designed for referring to single destination in appmap. 

So if users want to see links for single publisher and multiple subscribers on appmap, they should not set `messaging.url` and `net.peer.*` to keep Target empty.

If users want to see the link to dependency pubsub system regardless of topic, they can set `messaging.url` pr `net.peer.*`

So, the current change (referring to messaging.url and net.peer.*) is better fit for appinsight appmap. Apologies for the wrong suggestion. 

Once you fixed the unit-test, I will approve it. Thanks a lot for this work!

youngbupark replied:
Here is the screenshot. react-form publishes message to node-subscriber and python-subscriber via topic A and B. appmap shows the links between publisher and subscribers.

![image](https://user-images.githubusercontent.com/11863108/86538429-e130a200-beaa-11ea-8aa4-18929f121b92.png)

We can see the list of topic which node-subscriber subscribes.
![image](https://user-images.githubusercontent.com/11863108/86538450-fb6a8000-beaa-11ea-8833-2b9049c13d0e.png)

So the current change which doesn't refer to `messaging.destination` is better fit for app insight app map. If we use messaging.destination for Target, then we can see one link between react-form and node-subscriber or python-subscriber.

youngbupark reviewed (commented):
Thanks for changing all. I left one comment. other than that, lgtm. 

(seems like unti-test is broken in check-build)

pcwiese commented:
The failing test is from a different project and is unrelated. I'm assuming the test is a bit flaky. I just changed the commit message and force pushed to re-run the build-publish pipeline

pcwiese commented:
Can someone explain to me how this file  receiver/carbonreceiver/transport/tcp_server.go is showing up in my code coverage diff?

pcwiese committed:
Update Azure Monitor Exporter to use the new internal traces format

Description:
This change brings the Azure Monitor trace exporter up to date with
the latest OTEL semantic specifications for HTTP, RPC, Database, and
Messaging Spans. It does not implement any mappings for FaaS Spans.

This is a complete refactor to reduce code duplication and increase readability.

Testing:

 I ran Unit Tests locally.
 Tested E2E using a recent version of the OpenTelemetry dotnet SDK with matching otel proto files emitting HTTP server and client Spans.
Documentation:
An update to the README will follow in a separate PR from a colleague.

pcwiese commented:
Last change was a rebase from master only.

dmitryax commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -0,0 +1,679 @@
+
+	// if _, exists := attributeMap.Get(conventions.AttributeFaaSTrigger); exists {
+	// 	return faasSpanType
+	// }
```
Doesn't need commented right? I see that `faasSpanType` is handled in `spanToEnvelope`

pcwiese replied:
correct. uncommented.

dmitryax reviewed (approved):
Overall looks good to me, just a minor comment

pjanotti reviewed (approved):
LGTM
